lib/services/security/users.js

```diff
@@ -228,20 +243,27 @@ function __prepareUserForUpsert(user) {
 
 function __upsertUser(user) {
   return new Promise((resolve, reject) => {
+    let permissions = this.utilsService.clone(user.permissions);
     this.__prepareUserForUpsert(user).then(prepared => {
+      delete prepared.permissions; //Don't store in DB
       this.dataService.upsert(
         '/_SYSTEM/_SECURITY/_USER/' + prepared.username,
         prepared,
         { merge: true },
-        (e, result) => {
+        async (e, result) => {
           if (e) return reject(e);
           this.log.info(`user upserted: ${prepared.username}`);
           var upserted = this.securityService.serialize('user', result);
+          upserted.permissions = permissions || {};
+          if (permissions)
+            await this.permissionManager.upsertMultiplePermissions(upserted.username, permissions);
           this.securityService.dataChanged(
             CONSTANTS.SECURITY_DIRECTORY_EVENTS.UPSERT_USER,
             upserted,
             user,
             () => {
+              if (!permissions) return resolve(upserted);
```

Comment:
pls avoid mixing async/await and callback-driven styles within the same function.

some useful comments here on mixing async/await and callbacks:

https://stackoverflow.com/questions/62037166/can-i-mix-callbacks-and-async-await-patterns-in-nodejs

I'd like us to consider a rule where we don't mix these styles
it's easy enough to use util.promisify in an async function if relying on callback-driven libs

Reply:
this will also make it more feasible and comprehensible to wrap the entire async callback function in a try/catch statement to remove the possibility of unhandled promise rejections

Reply:
probably best to rewrite the whole __upsertUser function as an async function

Reply:
I'm also thinking about rewriting that method, as an async/await instead of raw promisifed function though

Reply:
even better